url27a7ae3508_serkan_ozal_ocean_of_memories_tgz-pJ8-com_zeroturnaround_rebellabs_oceanofmemories_article1_objectlayout_ObjectMemoryLayoutDemoJ8/src/com/zeroturnaround/rebellabs/oceanofmemories/common/util/JvmUtil.java

```diff
@@ -0,0 +1,2125 @@
+            this.compressedRef = true;
+            this.compressRefShift = shift;
+            this.compressRefBase = 0;
+        }
+
+        public VMOptions(String name, int shift, long base) {
+            this.name = name;
+            this.referenceSize = SIZE_32_BIT;
+            this.objectAlignment = guessAlignment(this.referenceSize) << shift;
+            this.compressedRef = true;
+            this.compressRefShift = shift;
+            this.compressRefBase = base;
+        }
+
+        public long toNativeAddress(long address) {
+            if (compressedRef) {
+                return compressRefBase + (address << compressRefShift);
+            } else {
+                return address;
+            }
+        }
+
+        public long toJvmAddress(long address) {
+            if (compressedRef) {
+                return (address >> compressRefShift) - compressRefBase;
+            } else {
+                return address;
+            }
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public boolean isCompressedRef() {
+            return compressedRef;
+        }
+
+        public int getCompressRefShift() {
+            return compressRefShift;
+        }
+
+        public long getCompressRefBase() {
+            return compressRefBase;
+        }
+
+        public int getObjectAlignment() {
+            return objectAlignment;
+        }
+
+        public int getReferenceSize() {
+            return referenceSize;
+        }
+
+    }
+
+    @SuppressWarnings("unused")
+    private static class CompressedOopsClass {
+
+        public Object obj1;
+        public Object obj2;
+
+    }
+
+    @SuppressWarnings("unused")
+    private static class HeaderClass {
+
+        public boolean b1;
+
+    }
+
+    private static class MyObject1 {
+
+    }
+
+    @SuppressWarnings("unused")
+    private static class MyObject2 {
+
+        private boolean b1;
+
+    }
+
+    @SuppressWarnings("unused")
+    private static class MyObject3 {
+
+        private int i1;
+
+    }
+
+    private static class Node {
+
+        private String name;
+        private List<Node> children;
+
+        private long shallowSize;
+        private long deepSize;
+
+        public Node(String name, Object delegate) {
+            this.name = name;
+
+            if (delegate != null) {
+                shallowSize = JvmUtil.shallowSizeOf(delegate);
+                deepSize = shallowSize;
+            }
+        }
+
+        private void addChild(Node node) {
+            if (children == null) {
+                children = new ArrayList<Node>();
+            }
+            children.add(node);
+            deepSize += node.deepSize;
+        }
+
+        public static Node create(Object delegate) {
+            return create("root", delegate,
+                    new IdentityHashMap<Object, Integer>());
+        }
+
+        public static Node create(String prefix, Object delegate,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.